### PR TITLE
Add Virtual Level Plugin

### PR DESCRIPTION
**Pull request by Joshua-F**
- Toggle for virtual level on skill tab
- Toggle for virtual total level on skill tab
- Toggle for virtual level on Xp Globes
- Toggle for virtual level on Hiscores plugin

Virtual level on skilltab toggle
Virtual level on Xp Globe toggle
(Ignore actions left, used `::addxp` to set the xp so I "gained" 13m xp in 1 action)

### CODE REVIEW

**Comment by SoyChai** [comment]
Very smooth, and I like how the total level is untouched by default due to the confusion that might cause. Amazing work.

File: runelite-client/src/main/java/net/runelite/client/plugins/virtuallevels/VirtualLevelsConfig.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	{
+		return true;
+	}
+
+	@ConfigItem(
+		position = 3,
+		keyName = "hiscoresPanel",
+		name = "Enable virtual levels on Hiscores",
+		description = "Configures if the hiscores panel shows virtual levels."
+	)
```
Comment (SoyChai) [inline runelite-client/src/main/java/net/runelite/client/plugins/virtuallevels/VirtualLevelsConfig.java]: We should capitalise this as HiScores to keep the client text consistent.

**Comment by Joshua-F** [comment]
Unsure why tests are crashing now.. Tried reverting locally and it still gives the same exception.

**Comment by LeviSchuck** [comment]
Have you tried to rebase on upstream master? It looks like it can't even load the base classes.

**Comment by Joshua-F** [comment]
The tests fail relating to trying to require a plugin as a dependency that happens to be disabled. `HiscorePlugin` depends on `VirtualLevelPlugin` and `VirtualLevelPlugin` is set to not load when outdated, which the tests force. This causes `HiscorePlugin` to cause the crash since it's attempting to depend on a plugin that isn't loaded. Adam talked about having optional plugins which might fix this problem, along with removing `VirtualLevelPlugin#isEnabled()`.

**Comment by LeviSchuck** [comment]
Looked at Guice, injections, providers and all that.

If you have something dynamic like this, and references are final, I think what you need is a provider that fulfills the interface, but what it gives is a a proxy object. The proxy object would route calls to the plugin when it is on, and when it is not it gives a stub implementation. 

This is looking more like a service, or at least, the idea of a service, than a plugin. Not that service is  a realized concept in RuneLite. 

A service can be available, or not.
Plugins depending on plugins seems icky to me.

**Comment by RomeuG** [comment]
How about a configuration setting to enable the displaying of the levels like 110/110 instead of 99/110?

**Comment by Joshua-F** [comment]
> How about a configuration setting to enable the displaying of the levels like 110/110 instead of 99/110?

I was orig going to do that by default but we (the few others I was talking about this with) couldn't come up with a clean way of handling boosts. Suggestions would be nice. Also, this is blocked by something that I'm not too sure how to do so it might be awhile before this is picked up.

**Comment by SoyChai** [comment]
I think a 110/110 approach would sacrifice usability for aesthetics, because the boost level would no longer make any sense.

**Comment by Joshua-F** [comment]
Could possibly do it how RS3 does, which is showing 110/110 if there were no boosts or anything applied. Otherwise it would show x/99. To me it's just confusing so I don't really know the solution to it.

**Comment by SoyChai** [comment]
...I actually like that approach a lot. Maybe we *could* display it the same way?

**Comment by Joshua-F** [comment]
Shouldn't be too hard to do. I'll see about rebasing this this weekend maybe and maybe talking to someone about how to avoid the issue I previously ran into. I think the solution Adam brought up was having optional plugins since this PR made it so a few plugins depend on it but it breaks if this plugin is disabled due to not running on outdated client for example.

**Pull request closed without merging**